### PR TITLE
Update Covers to handle AttributeRestrictions

### LINKED ISSUE #13312: Remove `PolicyRule`s with deprecated `AttributeRestrictions`
Once #13256 is merged, a rule with `AttributeRestrictions` will never match a `DefaultAuthorizationAttributes`.  These rules can be safely removed.

- [ ] `pkg/cmd/server/bootstrappolicy/policy.go`
- [ ] `pkg/authorization/authorizer/scope/converter.go`

### PR DESCRIPTION
**Pull request by enj**
Fixes #13312 

Needed to finish #13334 

cc @liggitt 

[test]

Signed-off-by: Monis Khan <mkhan@redhat.com>

### CODE REVIEW

**Comment by deads2k** [comment]
looks reasonable.

**Comment by liggitt** [comment]
did we remove all rules with attribute restrictions from bootstrap policy already?

**Comment by deads2k** [comment]
> did we remove all rules with attribute restrictions from bootstrap policy already?

There was a todo somewhere.  @enj this is a reasonable pull to do it in.

**Comment by liggitt** [comment]
it has to be done here, otherwise reconciliation will keep re-adding those rules, since they'll show up as uncovered

**Comment by enj** [comment]
> it has to be done here, otherwise reconciliation will keep re-adding those rules, since they'll show up as uncovered

The unit tests I wrote should fail if that was the case =\

**Comment by deads2k** [comment]
> it has to be done here, otherwise reconciliation will keep re-adding those rules, since they'll show up as uncovered

The rule doesn't do anything.  It's always covered because it's a rule that grants no permissions.  A rule with attributerestrictions is completely ignored by the authorizer.

**Comment by liggitt** [comment]
ah, do we breakdown both existing and expected rules?

**Comment by deads2k** [comment]
> ah, do we breakdown both existing and expected rules?

His update here doesn't allow a rule with attributerestrictions to cover anything and when it breaks down a rule with attributestrictions, it returns an empty slice of rules.

**Comment by enj** [comment]
@deads2k I removed all the rules with `AttributeRestrictions` that are not in test files.

**Comment by deads2k** [comment]
lgtm[merge]

**Comment by deads2k** [comment]
[merge]

**Comment by enj** [comment]
Flake #13426

**Comment by deads2k** [comment]
re[merge]

**Comment by deads2k** [comment]
re[test]

File: pkg/authorization/authorizer/scope/converter.go [hunk 1]
```diff
@@ -177,7 +177,6 @@ func (userEvaluator) ResolveRules(scope, namespace string, clusterPolicyGetter c
 	case UserAccessCheck:
 		return []authorizationapi.PolicyRule{
 			authorizationapi.NewRule("create").Groups(kauthorizationapi.GroupName).Resources("selfsubjectaccessreviews").RuleOrDie(),
-			{Verbs: sets.NewString("create"), APIGroups: []string{authorizationapi.GroupName, authorizationapi.LegacyGroupName}, Resources: sets.NewString("subjectaccessreviews", "localsubjectaccessreviews"), AttributeRestrictions: &authorizationapi.IsPersonalSubjectAccessReview{}},
```
Comment (smarterclayton) [inline pkg/authorization/authorizer/scope/converter.go]: This change broke the ability of a client to request a subject access review without scopes.  Scopes aren't in upstream Kube, and as a result, you can no longer use a token that only has "user:info user:check-access" scopes to do a SAR that tells you what the user can do.  Which breaks the ability to use this as a proxy.

If I can't do SAR and take scopes into account, then we are regressing functionality and shouldn't remove this until we have an equivalent behavior in Kube.  If there's a workaround I don't see it - couldn't find anything that allowed a separate channel for passing scopes.

Comment (smarterclayton) [inline pkg/authorization/authorizer/scope/converter.go]: to use as a proxy, without granting a token that gives you the ability to act as the full user.

Comment (liggitt) [inline pkg/authorization/authorizer/scope/converter.go]: Upstream has a dedicated resource for SelfSubjectAccessReview… no attribute restrictions needed

Comment (smarterclayton) [inline pkg/authorization/authorizer/scope/converter.go]: I can't pass empty scopes to check permissions on it like i can today.

**Comment by liggitt** [comment]
Ah. All the existing openshift APIs should still be usable and work identically. If they don't, that's a bug we have to fix. 

The goal of this was not to change that but to remove dependence on AttributeRestrictions. The special case of IsPersonalSubjectAccessReview got shifted from being embedded in the role attribute restrictions to being treated by the authorizer as a SelfSubjectAccessReview resource for authorization purposes. That should all be internal impl details… if anything leaked or broke, we'll need to resolve it.

**Comment by smarterclayton** [comment]
I think the gap is that a user:check-access user should be able to do the
equivalent of

POST /apis/authorization.openshift.io/subjectaccessreviews
{"scopes": [], "namespace": "foo", "resource": "pods", "verb": "get"}

as themselves.  Without an OpenShift "selfsubjectaccessreview", or the
authorizer allowing "subjectaccessreview" only for user unspecified
(meaning current user), there's no way to make the previous call.

**Comment by deads2k** [comment]
This changed the way the permission is expressed, but did not change the way the API works.  The openshift self-SAR endpoint is still available and accessible.  The permission used for the check is the kube self-sar, not the openshift self-sar, but the old endpoint is still there.

**Comment by deads2k** [comment]
> I think the gap is that a user:check-access user should be able to do the
> equivalent of

This is something that the upstream self-SAR can't do, but this pull did not remove the ability to use the old openshift API endpoint (or it wasn't intended to at any rate).  The openshift endpoint checks the kube self-SAR permission.

**Comment by smarterclayton** [comment]
The openshift endpoint isn't in the check-access scope, which this PR
removed on this line.  Or at least, I was unable to run it and got a
scope denial error.

**Comment by smarterclayton** [comment]
Fixed in #13715

**Pull request merged**